### PR TITLE
STL headers should require C++ and forbid C

### LINKED ISSUE #1141: STL headers should require C++ and forbid C
Until fairly recently, the STL's build contained C translation units, so some of our internal headers had to tolerate being compiled as C. Now that our build contains only C++ TUs, we should consider changing our headers to require C++. (This was prompted by discovering that an internal team had been including `<cassert>` in C mode. They've fixed their code.)

iso646.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/iso646.h#L6-L13

This header is special. It's a C header that lived in the STL's directory and (mistakenly?) wasn't moved to the UCRT. I believe we should:
* Stop including `<yvals_core.h>`
* Replace `#if _STL_COMPILER_PREPROCESSOR` with equivalent machinery
* Keep the `__cplusplus` guard
* ~~Investigate moving this to the UCRT properly~~ (Edit: not worth the effort)

xkeycheck.h
---
https://github.com/microsoft/STL/blob/e745bad3b1d05b5b19ec652d68abb37865ffa454/stl/inc/xkeycheck.h#L6-L15

On 2019-08-21, Microsoft-internal MSVC-PR-198458 changed this header to have no effect in C mode.

I think we should:

* Remove the `__cplusplus` guard
* Rely on `<yvals_core.h>` enforcing C++ mode

yvals_core.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals_core.h#L1237-L1247

This is the STL's central internal header. I think we should:

* Remove the C definitions here
* Add an `#error` elsewhere in the file, probably early, forbidding it from being compiled as C (like how we forbid old compilers)

yvals.h
---
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals.h#L133-L153
https://github.com/microsoft/STL/blob/b74b6188b233cce0647950d7fd6bbbb3bd3744f0/stl/inc/yvals.h#L325-L501

This is the STL's "runtime support" internal header. I think we should:
* Remove the `__cplusplus` guards here
* Rely on `<yvals_core.h>` enforcing C++ mode

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #1141

### CODE REVIEW

File: stl/inc/iso646.h [hunk 1]
```diff
@@ -6,7 +6,18 @@
 #pragma once
 #ifndef _ISO646
 #define _ISO646
-#include <yvals_core.h>
+
+#ifndef _STL_COMPILER_PREPROCESSOR
+// All STL headers avoid exposing their contents when included by various
+// non-C++-compiler tools to avoid breaking builds when we use newer language
+// features in the headers than such tools understand.
+#if defined(RC_INVOKED) || defined(Q_MOC_RUN) || defined(__midl)
```
[the remark this answers is not in the document]
Comment (fsb4000) [reply stl/inc/iso646.h:14]: I don't know...

I just copy-pasted from there:
https://github.com/microsoft/STL/blob/f75c7f596c7b491168fefb5eff5164ab7ae36867/stl/inc/yvals_core.h#L9-L18

I would not want it to be any different. But I will wait for more opinions...

[the remark this answers is not in the document]
Comment (cpplearner) [reply stl/inc/iso646.h:14]: It is undefined behavior if `defined` is generated as a result of macro expansion.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
There will be a stealth merge conflict with #2140, where `<stdatomic.h>` wants to emit a specific `#error` for C. I believe we could simply move that `#error` above `#include <yvals.h>`.

**Comment by StephanTLavavej** [comment]
@fsb4000 @CaseyCarter I've pushed a series of commits:

* Merged with `main`.
* Added a period to the error message, for consistency with the existing messages: https://github.com/microsoft/STL/blob/78ff4618d4fa43eeb4bb714d76f27d88cf11a1b6/stl/inc/yvals_core.h#L585
* In `yvals_core.h`, moved the check within `_STL_COMPILER_PREPROCESSOR`. We want to continue to expand to nothing for known non-compiler tools.
* Change `iso646.h` to avoid defining `_STL_COMPILER_PREPROCESSOR` (which may be problematic for header unit scenarios where two header units try to emit the same macro independently) and instead just repeat the logic inline, with a bit of commenting.
* Resolve the stealth merge conflict in `stdatomic.h` by repeating the `_STL_COMPILER_PREPROCESSOR` logic inline, then emitting a specific error before the general one.

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
I have mirrored this to internal MSVC-PR-347845 (all by itself, as I believe it has a moderate risk of breaking stuff). Please notify me if any further changes are pushed.

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for improving the diagnostics here! :tada: :smile_cat: :rocket:

**Pull request merged**